src/Confluent.Kafka/ProducerBuilder.cs

```diff
@@ -20,6 +20,39 @@
 
 namespace Confluent.Kafka
 {
+    /// <summary>
+    ///     Calculate a partition number given a <paramref name="topic" />,
+    ///     <paramref name="partitionCount" /> and serialized <paramref name="keyData" />.
+    /// </summary>
+    /// <remarks>
+    ///     A partioner instance may be called in any thread at any time and
+    ///     may be called multiple times for the same message/key.
+    ///
+    ///     A partitioner:
+    ///     - MUST NOT call any method on the producer instance.
+    ///     - MUST NOT block or execute for prolonged periods of time.
+    ///     - MUST return a value between 0 and partitionCnt-1.
+    ///     - MUST NOT throw any exception.
+    /// </remarks>
+    /// <param name="topic">
+    ///     The topic.
+    /// </param>
+    /// <param name="partitionCount">
+    ///     The number of partitions in <paramref name="topic" />.
+    /// </param>
+    /// <param name="keyData">
+    ///     The serialized key data.
+    /// </param>
+    /// <param name="keyIsNull">
+    ///     Whether or not the key is null (distinguishes the null and empty case).
```

Comment:
It's too bad we can't utilize sticky partitioning when a custom partitioner is set.
We could add a new API, SetNonRandomPartitioner (and the analogue on the rdkafka side) to have a custom partitioner that is only passed non-NULL keys.

Reply:
i'd do that with another argument on `SetPartitioner` / `SetDefaultPartitioner` (a boolean). it can be added later as a non breaking change by giving it a default value.